### ISSUE TITLE
clang 3.6 + libc++ compilation error

### ISSUE THREAD

**Issue by tjhei** [dealii/dealii#1134]
installed from ubuntu packages, configured with "-stdlib=libc++":
```
FAILED: /usr/bin/clang++-3.6   -DDEBUG -DTBB_DO_ASSERT=1 -DTBB_USE_DEBUG -Isource/lac -Iinclude -I/scratch/deal-git/include -I/scratch/deal-git/bundled/tbb41_20130401oss/include -I/scratch/deal-git/bundled/umfpack/UMFPACK/Include -I/scratch/deal-git/bundled/umfpack/AMD/Include -I/scratch/deal-git/bundled/boost-1.56.0/include -I/scratch/deal-git/bundled/muparser_v2_2_4/include    -pedantic -fpic -Wall -Wextra -Wpointer-arith -Wwrite-strings -Wsynth -Wsign-compare -Wswitch -Wno-long-long -Wno-deprecated-declarations -Qunused-arguments -Wno-unsupported-friend -Wno-unused-parameter -Wno-unused-variable -Wno-c99-extensions -Wno-variadic-macros -Wno-c++11-extensions -std=c++14 -stdlib=libc++ -Wno-parentheses -Wno-unused-local-typedefs -O0 -ggdb -Wa,--compress-debug-sections -MMD -MT source/lac/CMakeFiles/obj_lac.debug.dir/sparse_vanka.cc.o -MF "source/lac/CMakeFiles/obj_lac.debug.dir/sparse_vanka.cc.o.d" -o source/lac/CMakeFiles/obj_lac.debug.dir/sparse_vanka.cc.o -c /scratch/deal-git/source/lac/sparse_vanka.cc
In file included from /scratch/deal-git/source/lac/sparse_vanka.cc:16:
In file included from /scratch/deal-git/include/deal.II/lac/sparse_vanka.templates.h:20:
In file included from /scratch/deal-git/include/deal.II/base/memory_consumption.h:20:
In file included from include/deal.II/base/config.h:339:
In file included from /scratch/deal-git/include/deal.II/base/numbers.h:22:
In file included from /usr/include/c++/v1/complex:244:
/usr/include/c++/v1/type_traits:433:83: error: no matching function for call to '__source'
    : public integral_constant<bool, sizeof(__libcpp_is_function_imp::__test<_Tp>(__libcpp_is_function_imp::__source<_Tp>())) == 1>
                                                                                  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/include/c++/v1/type_traits:438:14: note: in instantiation of template class 'std::__1::__libcpp_is_function<void (dealii::Vector<float> &, const dealii::Vector<float> &, const std::__1::vector<bool, std::__1::allocator<bool> > *) const, false>' requested here
    : public __libcpp_is_function<_Tp> {};
             ^
/usr/include/c++/v1/type_traits:443:97: note: in instantiation of template class 'std::__1::is_function<void (dealii::Vector<float> &, const dealii::Vector<float> &, const std::__1::vector<bool, std::__1::allocator<bool> > *) const>' requested here
template <class _Tp, class _Up> struct __libcpp_is_member_function_pointer<_Tp _Up::*> : public is_function<_Tp> {};
                                                                                                ^
/usr/include/c++/v1/type_traits:446:14: note: in instantiation of template class 'std::__1::__libcpp_is_member_function_pointer<void (dealii::SparseVanka<double>::*)(dealii::Vector<float> &, const dealii::Vector<float> &, const std::__1::vector<bool, std::__1::allocator<bool> > *) const>' requested here
    : public __libcpp_is_member_function_pointer<typename remove_cv<_Tp>::type> {};
             ^

....

/scratch/deal-git/include/deal.II/base/thread_management.h:1993:8: note: while substituting deduced template arguments into function template 'function' [with _Fp = std::__1::__bind<void (dealii::SparseVanka<double>::*&)(dealii::Vector<float> &, const dealii::Vector<float> &, const std::__1::vector<bool, std::__1::allocator<bool> > *) const, std::__1::reference_wrapper<const dealii::SparseVanka<double> >, std::__1::reference_wrapper<dealii::Vector<float> >, std::__1::reference_wrapper<const dealii::Vector<float> >, const std::__1::vector<bool, std::__1::allocator<bool> > *>]
      (std_cxx11::bind(fun_ptr, std_cxx11::cref(c),
       ^
/scratch/deal-git/include/deal.II/lac/sparse_vanka.templates.h:586:29: note: in instantiation of function template specialization 'dealii::Threads::new_thread<void, dealii::SparseVanka<double>, dealii::Vector<float> &, const dealii::Vector<float> &, const std::__1::vector<bool, std::__1::allocator<bool> > *>' requested here
        threads += Threads::new_thread (comp,
                            ^
```

This seems to be related: https://llvm.org/bugs/show_bug.cgi?id=20084

What do we do, disable cxx11 if we detect the bug?

**Comment by tjhei** [comment]
minimal failing test
```
#include <functional>
struct A {
    int foo() const {}
};
int main()
{
  A a;
  auto x=std::bind(&A::foo,a);
  x();
}
```

**Comment by tamiko** [comment]
Sounds good.

**Comment by bangerth** [comment]
This was fixed by #1135.
